### DISCUSSION TITLE
Setting Color Values + Line Follower

### DISCUSSION THREAD

**Discussion by TheWendyPower** [pybricks/support#967] [Q&A] [answered]
My overall goal is to get line following working. My first snag is coming in setting color values. By default Color.Black isn't recognized. 

```python
color_sensor = ColorSensor(Port.A)
color = color_sensor.color(surface=True)
color_true = color_sensor.hsv(surface=True)
Color.BLACK = Color(h=210, s=17, v=15)
my_colors = (Color.BLACK, Color.WHITE)
color_sensor.detectable_colors(my_colors)
```

But when I run 

```pyrhon
print(color,reflection)
```

I get back `Color.NONE 7`

At least with the reflection value I know the sensor is working. 

With the line following I will want to use reflected light. However, I want the robot to get to the black line first before starting the line following. I have used reflected light:

```python
def wait_for_black():
    while color_sensor.reflection() > 20:
        wait(10)
```

but it only works so-so. If I'm off just a little bit the robot fails to stop at the black line. I have played with that number. The current value has been the most consistent (using an FLL mat from last year). 

I'm sure the issue is something I'm doing wrong. I've spent the day on it and can't find where I'm messing things up.

**Comment by TheWendyPower** [comment, marked as answer]
My line follow code just wants to turn right:

```python
hub = PrimeHub()

#Drive Setings
left_drive = Motor(Port.C, Direction.COUNTERCLOCKWISE,reset_angle=True)
right_drive = Motor(Port.B, Direction.CLOCKWISE, reset_angle=True)
motor_pair = DriveBase(left_motor=left_drive, right_motor=right_drive, wheel_diameter=55.8, axle_track=112)

#Attachment Settings
lift = Motor(Port.D,Direction.CLOCKWISE, [36,36,12], reset_angle=False)

#Sensor Setting
color_sensor = ColorSensor(Port.A)
color = color_sensor.color(surface=True)
color_true = color_sensor.hsv(surface=True)
reflection = color_sensor.reflection()
Color.BLACK = Color(h=240, s=7, v=15)
my_colors = (Color.BLACK, Color.WHITE, Color.NONE)
color_sensor.detectable_colors(my_colors)
black = 6
white = 77
threshold = (black + white) / 2
PROPORTIONAL_GAIN = 1.5
deviation = reflection - threshold
turn_rate = PROPORTIONAL_GAIN * deviation

#Functions
def wait_for_black():
    while color_sensor.reflection() > 20:
        wait(10)

def drive_to_black(speed, trun):
    motor_pair.drive(speed,trun)
    wait_for_black()
    motor_pair.stop()

def line_follower(speed, distance):
    motor_pair.reset()
    while motor_pair.distance() <= distance * 10: #turn CM to MM
        motor_pair.drive(speed,turn_rate)
    motor_pair.stop()
```

**Reply by dlech** [reply]
There is nothing in the while loop to change the value of `turn_rate` so it will always drive in the same direction.

```python
reflection = color_sensor.reflection()
deviation = reflection - threshold
turn_rate = PROPORTIONAL_GAIN * deviation
```

should be inside the loop so that the value updates response to the changing sensor value

**Reply by TheWendyPower** [reply]
Thank you! I knew I was something simple I had been overlooking. I had deviation and turn_rate in the loop before, during my trial and errors but never reflection.

**Comment by TheWendyPower** [comment]
Thank you @dlech  for pointing out what I was missing in my line follow function. However, it won't go backward. Honestly, I can seem to get any of the motors (using a contentious run option) to run backward at all. That is both Drive Base and individual motors. Is there something else I'm missing? I admit I'm still in the process of learning python and most of my experience with it has been the version from Lego. 

I'm also still having the issue where the sensor is reading no color on black, even with it defined like above.

**Reply by dlech** [reply]
To make it go backwards, you can just give it a negative speed.

For the color sensor, I would recommend using one of the example programs like https://docs.pybricks.com/en/latest/pupdevices/colorsensor.html#measuring-color-and-reflection and just more the robot around by hand to get a feel for how the color sensor works. It could be that the color sensor is too far away to properly detect colors or something like that.

**Reply by TheWendyPower** [reply]
Maybe I need to just reinstall. I am giving it a negative speed or distance, but nothing happens. I can use:
`motor_pair.straight()` with a positive or negative and get the desired result. `motor_pair.drive()` with a positive speed gets me to go forward but a negative speed does nothing. I get the same thing on the individual motors. `left_drive.run()` will work with a positive number, but does nothing with a negative number. 

It will read other colors fine, the ones that are originally included anyway.

**Reply by TheWendyPower** [reply]
I think I got it figured out. The way I have my function written was one part of that. It also seems the drive forever portions would let you use them unless there is a way for them to eventually stop.

**Reply by laurensvalk** [reply]
> It also seems the drive forever portions would let you use them unless there is a way for them to eventually stop.

If the `run()` or `drive()` is the last thing in your program, the program ends right after that, so the motors stop right away.

Adding something like a `wait()` or anything else that takes some time, like waiting for a condition, should give you some time to see it move.

**Reply by TheWendyPower** [reply]
@laurensvalk it would be great if that was added to the documentation. I figured it was most likely something that I was doing wrong. I checked the documentation on motors multiple times. It doesn't mention you MUST have a wait or some other function to break the motor running "forever" or it jumps to the next line of code. For many that are more versed in micro python that is probably a given. For those of us coming from LEGO Python, where that wasn't a condition of using a continuous run function of a single motor or drive base, noting that would be helpful. 

I  first noticed the issue of trying to write a line squaring function. I've got one that works now, though it probably isn't very efficient. 

```
def line_square_right(speed,distance):
    drive_to_black(speed, 0, right_color)
    turn_degrees(360, False)
    wait_for_black(left_color)
    motor_pair.stop()
    drive_distance(distance, True)
    turn_degrees(-365,False)
    wait_for_black(right_color)
    motor_pair.stop()

def line_square_left(speed, distance):
    drive_to_black(speed, 0, left_color)
    turn_degrees(-360, False)
    wait_for_black(right_color)
    motor_pair.stop()
    drive_distance(distance, True)
    turn_degrees(365,False)
    wait_for_black(left_color)
    motor_pair.stop()
```

Having the turn rate slower would help on the line squaring, but I can play with that setting as well. As for the line follower, it looks like this:

```
def line_follower(speed, distance, sensor):
    motor_pair.reset()
    PROPORTIONAL_GAIN = 2
    while motor_pair.distance() != distance * 10: #turn CM to MM
        reflection = sensor.reflection()
        deviation = reflection - threshold
        turn_rate = PROPORTIONAL_GAIN * deviation
        motor_pair.drive(speed,turn_rate)
    motor_pair.stop()
```

Going from less than or equal to not equal means my function can go backward. It just doesn't perform well backward, resulting in oversteering. At one point I had a second line follow just for reverse. There I adjusted the proportional gain and multiplied the turn rate by -1. When driving my car backward I need to turn the steering wheel in the opposite direction to get it to go the way I want. This should take my resulting turn rate and make them opposite from the forward turn rate. However, I still ended up with over-corrections and poor line following in reverse. I'll play with it some more. However, `straight()` works so well, and our backup distances will be much shorter that is probably the route we will go instead of the line following backward.

**Reply by laurensvalk** [reply]
Thanks, this would certainly be good to document indeed. Stopping when the program ends was the LEGO standard for about 20 years but it changed with SPIKE.

**Reply by TheWendyPower** [reply]
And I have only used SPIKE. We do have a couple of EV3 hubs that were recently given to us with their motors/sensors/parts. However, with 3 SPIKE hubs and 1 SPIKE essential hub we haven't messed with the EV3, just used their parts.

**Reply by laurensvalk** [reply]
Just a note for future reference:

Equality / inequality conditions like these may be missed in some circumstances

```python
motor_pair.distance() != distance * 10
```

There is no guarantee that the distance measurement will always pass through 80, for example. It could jump from 79 to 81, which might not be what you want. (And the issue may not happen until competition day)

You could make it more reliable by relaxing the check by, say, 5 millimeters in both directions:

```python
not (distance * 10 - 5 < motor_pair.distance() < distance * 10 + 5)
```

Or

```python
abs(distance * 10 - motor_pair.distance()) > 5
```

**Reply by laurensvalk** [reply]
> And I have only used SPIKE. We do have a couple of EV3 hubs that were recently given to us with their motors/sensors/parts. However, with 3 SPIKE hubs and 1 SPIKE essential hub we haven't messed with the EV3, just used their parts.

Yes, we probably got used to it so much that we didn’t think to document it. Thanks for pointing it out!

Please do share things like this as you run into them, we really appreciate it.

**Reply by TheWendyPower** [reply]
The whole point of using a line-follow function is consistency. Any team that is scoring big or has a robot that is getting around the same points every run is using some sort of line-follow function. Thank you for that suggestion.

**Comment by laurensvalk** [comment]
> I'm also still having the issue where the sensor is reading no color on black, even with it defined like above.

```python
color_sensor = ColorSensor(Port.A)
color = color_sensor.color(surface=True)
color_true = color_sensor.hsv(surface=True)
Color.BLACK = Color(h=210, s=17, v=15)
my_colors = (Color.BLACK, Color.WHITE)
color_sensor.detectable_colors(my_colors)
```
> But when I run
```python
print(color,reflection)
```
> I get back `Color.NONE 7`

Could this be a similar issue as the one in the line following program?

The snippet above prints an old measurement, taken before the calibrated values were set.

What happens if you try this?

```python
print(color_sensor.color(), color_sensor.reflection())
```

**Reply by TheWendyPower** [reply]
I ran `print(color_sensor.color(), color_sensor.reflection())` and got the correct color readout. YAY!!!

After that I moved the variables for `color()`, `reflection()`, and `hsv()` everything was working how it should. The lesson, pay more attention to the order of items in the code. While there has been a learning curve for my novice self, I do appreciate the way it is making us write better code.

**Reply by laurensvalk** [reply]
Glad to hear it worked!

**Comment by coxmin1** [comment]
Hey I am trying to convert this code to a 2 sensor line follower. Any ideas/suggestions/tips?

**Reply by ggramlich** [reply]
Yes. Instead of

`deviation = reflection - threshold`

simply use

`deviation = left_sensor.reflection() - right_sensor.reflection()`

You probably need to lower PROPORTIONAL_GAIN as the difference here is about twice as much as the difference to the threshold.

**Reply by coxmin1** [reply]
Damn that was a fast reply. Thank you I will put it to practice

**Reply by laurensvalk** [reply]
And how should the color sensors be mounted on the robot?

**Reply by coxmin1** [reply]
I have them mounted one stud apart. I'm working with the WRO senior high mat and the black line is about 2cm or 0.8in in width

**Reply by laurensvalk** [reply]
Awesome - just wanted to make this question/answer self contained when someone finds it later. Thanks!
